src/main/scala/mesosphere/util/state/zk/ZKStore.scala

```diff
@@ -0,0 +1,149 @@
+package mesosphere.util.state.zk
+
+import java.util.UUID
+
+import com.fasterxml.uuid.impl.UUIDUtil
+import com.google.protobuf.{ InvalidProtocolBufferException, ByteString }
+import com.twitter.util.{ Future => TWFuture }
+import com.twitter.zk.{ ZNode, ZkClient }
+import mesosphere.marathon.{ Protos, StoreCommandFailedException }
+import mesosphere.util.ThreadPoolContext
+import mesosphere.util.state.zk.ZKStore._
+import mesosphere.util.state.{ PersistentEntity, PersistentStore }
+import org.apache.log4j.Logger
+import org.apache.zookeeper.KeeperException
+import org.apache.zookeeper.KeeperException.NoNodeException
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ Await, Future, Promise }
+
+class ZKStore(val client: ZkClient, rootNode: ZNode) extends PersistentStore {
+
+  private[this] val log = Logger.getLogger(getClass)
+  private[this] implicit val ec = ThreadPoolContext.context
+
+  val root = createPath(rootNode)
+
+  /**
+    * Fetch data and return entity.
+    * The entity is returned also if it is not found in zk, since it is needed for the store operation.
+    */
+  override def load(key: ID): Future[Option[ZKEntity]] = {
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
+    node.getData().asScala
+      .map{ data => Some(ZKEntity(node, ZKData(data.bytes), Some(data.stat.getVersion))) }
+      .recover { case ex: NoNodeException => None }
+      .recover(exceptionTransform(s"Could not load key $key"))
+  }
+
+  override def create(key: ID, content: Array[Byte]): Future[ZKEntity] = {
+    val node = root(key)
+    require(node.parent == root, "Nested paths are not supported!")
+    val data = ZKData(key, UUID.randomUUID(), content)
+    node.create(data.toProto.toByteArray).asScala
+      .map(n => ZKEntity(n, data, Some(0))) //first version after create is 0
+      .recover(exceptionTransform(s"Can not create entity $key"))
+  }
+
+  /**
+    * This will store a previously fetched entity.
+    * The entity will be either created or updated, depending on the read state.
+    * @return Some value, if the store operation is successful otherwise None
+    */
+  override def save(entity: PersistentEntity): Future[ZKEntity] = {
+    val zk = zkEntity(entity)
+    def update(version: Int): Future[ZKEntity] = {
+      zk.node.setData(zk.data.toProto.toByteArray, version).asScala
+        .map { data => zk.copy(version = Some(data.stat.getVersion)) }
+        .recover(exceptionTransform(s"Can not update entity $entity"))
+    }
+    def create(): Future[ZKEntity] = {
+      zk.node.create(zk.data.toProto.toByteArray).asScala
+        .flatMap { _.getData().map{ data => zk.copy(version = Some(data.stat.getVersion)) }.asScala }
+        .recover(exceptionTransform(s"Can not update entity $entity"))
+    }
+    zk.version.map(update).getOrElse(create())
```

Comment:
The decision whether to save(= update) or to create (= insert) is now actually handled in `MarathonStore.modify`. Thus this code can go. Or did I misunderstand something?